src/main/java/org/openstreetmap/atlas/checks/validation/intersections/AtGradeSignPostCheck.java

```diff
@@ -0,0 +1,634 @@
+     * @return an optional {@link CheckFlag} object that
+     */
+    @Override
+    protected Optional<CheckFlag> flag(final AtlasObject object)
+    {
+        final Node intersectingNode = (Node) object;
+        // Filter and sort all in edges that have valid highway types
+        final List<Edge> inEdges = intersectingNode.inEdges().stream()
+                .filter(this::isValidIntersectingEdge)
+                .sorted(Comparator.comparingLong(AtlasObject::getIdentifier))
+                .collect(Collectors.toList());
+        // Filter all out edges that have valid highway types
+        final Set<Edge> outEdges = intersectingNode.outEdges().stream()
+                .filter(this::isValidIntersectingEdge).collect(Collectors.toSet());
+        // Terminate if there isn't at least one inEdge or two out edges with valid highway types
+        if (inEdges.isEmpty() || outEdges.size() < 2)
+        {
+            return Optional.empty();
+        }
+        // For each inEdge, get the list of potentially matching out edges
+        // Matching out edges are based on z level and highway type.
+        // For each inEdge, store the inEdge and corresponding outEdges in inEdgeToOutEdgeMap
+        // If any of the out edge is a roundabout edge, store the roundabout edges and the inEdge in
+        // roundAboutInEdgeToOutEdgeMap
+        final Map<String, Map<AtlasEntity, Set<AtlasEntity>>> mapOfMatchingInAndOutEdges = this
+                .populateInEdgeToOutEdgeMaps(inEdges, outEdges);
+        final Map<AtlasEntity, Set<AtlasEntity>> inEdgeToOutEdgeMap = mapOfMatchingInAndOutEdges
+                .getOrDefault(NON_ROUNDABOUT_INTERSECTION_MAP, null);
+        final Map<AtlasEntity, Set<AtlasEntity>> roundAboutInEdgeToOutEdgeMap = mapOfMatchingInAndOutEdges
+                .getOrDefault(ROUNDABOUT_INTERSECTION_MAP, null);
+        // If there are no valid intersection, return Optional.empty()
+        if ((inEdgeToOutEdgeMap == null || inEdgeToOutEdgeMap.isEmpty())
+                && (roundAboutInEdgeToOutEdgeMap == null || roundAboutInEdgeToOutEdgeMap.isEmpty()))
+        {
+            return Optional.empty();
+        }
+        // Collect all destination sign relations, the node is a member of
+        final Optional<Set<Relation>> destinationSignRelations = this
+                .getParentDestinationSignRelations(intersectingNode);
+        final FlaggedIntersection flaggedIntersection = destinationSignRelations.isEmpty()
+                ? this.getFlaggedIntersection(roundAboutInEdgeToOutEdgeMap, inEdgeToOutEdgeMap)
+                : this.getIntersectionsWithIncompleteDestinationSignRelation(
+                        roundAboutInEdgeToOutEdgeMap, inEdgeToOutEdgeMap, intersectingNode,
+                        destinationSignRelations.get());
+        final int instructionIndex = flaggedIntersection.getInstructionIndex();
+        final Set<AtlasEntity> entitiesToBeFlagged = flaggedIntersection.getFlaggedItems();
```

Comment:
Are we verifying that getFlaggedItems() is not null or not empty. Because, setOfFlaggedItems variable can be null.

Reply:
he instructionIndex is updated only if there is some item to be flagged and the line 217 ensures that Optional.empty() is returned if there are no flaggedItems.

Reply:
Yes, you are right. instructionIndex will be -1 if no items to be flagged and also getFlaggedItems() can be null because setOfFlaggedItems is not set anywhere. So, please move the if condition above this line.

Reply:
flaggedIntersection.getFlaggedItems() wouldn't be null in any case. It would be an empty set. But if instructionIndex =-1, flaggedIntersection.getFlaggedItems() need not be called and so I moved the line as requested.